### ISSUE TITLE
IPTables Redirect rule causes tcp connection timeouts with multiple ports

### ISSUE THREAD

**Issue by klarose** [istio/istio#38982]
### Symptoms

A virtual service for tcp traffic occasionally fails to connect to a kubernetes pod with multiple exposed ports. Occasionally tcp connections from our istio ingress to a pod with an istio sidecar fail with a UF, URX flag. 
Adding retries to our affected tcp proxies removed the connectivity problem from the perspective of the client. However, establishing the connection still took > 1.5s because the ingress would need to wait for the connection timeout to detect the failure, which isn't an acceptable perf hit when it is otherwise avoidable.

Digging in, we found that the failures were caused by the pod dropping the SYN packet and replying to a *different* connection with an [rfc5961](https://datatracker.ietf.org/doc/html/rfc5961) Challenge Ack. In particular, we saw that the istio ingress would send a SYN to port 5002. The destination
pod did not respond to that SYN. Instead, it would send a Challenge ACK to the same ip and port, but from port 5000. After some thought, it became apparent that the iptables redirect rules mapping all connections to the same destination port was
leading to collisions in the pod's connection table, and thus timeouts on the colliding connections.

We could see TcpExtTCPChallengeACK in the linux nstat tool incrementing when this occurs, which aligns with our theory.

### Setup
- One service with port 5000 and port 5002. 
- Two virtual services: One forwarding http from host X to port 5000. The other sni forwarding from host Y to port 5002.

### Analysis

The istio init container configures some iptables rules which ensure that the istio sidecar receives all inbound traffic. It does this by redirecting all connections to a specific destination port (say 15006). You can see this by viewing
the configured listeners:

```
0.0.0.0       15006 Trans: raw_buffer; Addr: *:5002                                                                 Cluster: inbound|5002||
0.0.0.0       15006 Trans: tls; Addr: *:5002                                                                        Cluster: inbound|5002||
0.0.0.0       15006 Trans: tls; App: istio,istio-peer-exchange,istio-http/1.0,istio-http/1.1,istio-h2; Addr: *:5002 Cluster: inbound|5002||
0.0.0.0       15006 Trans: raw_buffer; Addr: *:5000                                                                 Cluster: inbound|5000||
0.0.0.0       15006 Trans: tls; App: istio,istio-peer-exchange,istio-http/1.0,istio-http/1.1,istio-h2; Addr: *:5000 Cluster: inbound|5000||
```

This means that when a request arrives at the pod to port X, its destination port is rewritten to port 15006 for all ports (aside from some exclusions specific to the istio infrastructure). This is what introduces the problem.

It is perfectly possible for the following entries to exist in the connection table simultaneously: 
|sip|sport|dip|dport|
|---|-----|---|-----|
A|X|B|Y
A|X|B|Z

Note that the source port and source IP are the same. This is okay, becuase the four-tuple is what uniquely identifies them for a given protocol (e.g. tcp). However, this is not acceptable:

|sip|sport|dip|dport|
|---|-----|---|-----|
A|X|B|15006
A|X|B|15006

You cannot have two entries with the same four-tuple.

This is what we see happening. A long-lived connection to 5000 (A, X, B, 5000) collides with a new connection (A, X, B, 5002). After the rewrite, the kernel sees two
entries for A|X|B|15006, so it sends a challenge ack to A, X from port 5000, discarding the SYN.

It is worth noting that as the number of concurrent connections between A and B increases, so does the chance of a collision, since the chance of a port being reused increases.

#### Example

For a concrete example, consider the following setup and sequence of events.

10.1.134.168 is my ingress IP
10.1.134.72  is my pod IP.

My pod exposes two services: one on port 5000, one on port 5002.

0: New connection from (10.1.134.168, 38208) arrives at (10.1.134.72 , 5000).
1: iptables redirects to port 15006. Connection is now (10.1.134.168, 38208, 10.1.134.72 , 15006)
2. Traffic runs between  (10.1.134.168, 38208)->(10.1.134.72 , 15006) all fine and dandy. For a while.
3. Some time later, new a SYN from (10.1.134.168, 38208) arrives for (10.1.134.72 , 5002)
4. Iptables redirects to port 15006.  Connection is now (10.1.134.168, 38208, 10.1.134.72 , 15006)
5. Kernel sees that there is an existing connection for (10.1.134.168, 38208, 10.1.134.72, 15006). It sends the challenge ack to the source of the original connection.
6. The original connection is mapped back to (10.1.134.168, 38208, 10.1.134.72, 5000)
7. The challenge ack is sent to (10.1.134.168, 38208) from port 5000.
8. The SYN is dropped.
9. ~1.5s later the connection attempt from 10.1.134.168 times out after steps 0 through 8 repeat due to the SYN retransmitting.

You can see this in my attached capture: [synchallenge.pcap.zip](https://github.com/istio/istio/files/8711572/synchallenge.pcap.zip)
 
There is onging traffic between 38208 and 5000. Packet 17 shows the SYN. Packet 18 is the Challenge ACK. The stack
retries the SYN at packet 23, and again you see a challenge ACK.

### Workaround
Exclude port 5002 from the redirect in the pod template 
annotations, and set up an explicit sidecar object for this. E.g:

```
apiVersion: networking.istio.io/v1beta1
kind: Sidecar
metadata:
  name: disable-redirect
  namespace: my-namespace
spec:
  workloadSelector:
    labels:
      app: my-appp
  ingress:
    - port:
        number: 5002
        protocol: TCP
      defaultEndpoint: 127.0.0.1:6002
      captureMode: NONE
```

### What I expect to happen

I do not expect connections to randomly fail between the ingress and sidecar when my cluster is otherwise stable, and my system isn't really under load. Once in a while is fine; retries can handle that. However, I found a failure rate of around 0.1% in one of my tests which is far too high. Further, this smells like it could lead to other issues than just basic connectivity problems once in a while. Fundamentally, mapping multiple ports many to one is just asking for trouble.

I do not expect to need to apply manual workarounds in the form of Sidecar config + injection overrides in the pod template when the configuration I have deployed seems perfectly in line with the configuration model -- i.e. it's not some weird corner case.

### Reproduction

I'll try to come up with a simpler scenario, but basically a pod with a sidecar and two ports exposed. A few long lived connections on one port, and a series of 'cycling' connections on the other. The easiest way to get that is to have one port be http (so the connection pool is in play), and the other be tcp or tls.

Keep a few http requests going to the http endpoint so the connection pool stays primed. Then, simply send a continuous set of new connections into the ingress targeting the tls endpoint. Some of them will fail to connect with the default configuration (which has no tcp retries). If tcp retries are enabled, you'll see the connections that would have failed take much longer to establish.

### Version

```prose
$ istioctl version
client version: 1.10.6
control plane version: 1.10.6
data plane version: 1.10.6 (126 proxies)

$ istioctl version --short
client version: 1.10.6
control plane version: 1.10.6
data plane version: 1.10.6 (126 proxies)

I also reproduced the issue and workaround on Istio 1.13.3.
```

### bug report
```
istioctl bug-report

Target cluster context: XXXXXX

Running with the following config: 

istio-namespace: istio-system
full-secrets: false
timeout (mins): 30
include: {  }
exclude: { Namespaces: kube-system, kube-public, kube-node-lease, local-path-storage } AND { Namespaces: kube-system, kube-public, kube-node-lease, local-path-storage }
end-time: 2022-05-17 17:06:18.541145161 -0400 EDT


The following Istio control plane revisions/versions were found in the cluster:
Revision default:
&version.MeshInfo{
    {
        Component: "pilot",
        Info:      version.BuildInfo{Version:"1.10.6", GitRevision:"fd053c6165d21105d66dac6e3d0649db2dde5b86", GolangVersion:"", BuildStatus:"Clean", GitTag:"1.10.6"},
    },
    {
        Component: "pilot",
        Info:      version.BuildInfo{Version:"1.10.6", GitRevision:"fd053c6165d21105d66dac6e3d0649db2dde5b86", GolangVersion:"", BuildStatus:"Clean", GitTag:"1.10.6"},
    },
}

The following proxy revisions/versions were found in the cluster:
Revision default: Versions {1.10.6}
```

I cannot include any more information as it contains too many sensitive details. Let me know if there is anything in particular which would be useful and I can try to gather it.

### Environments tested on

- gke 1.21, istio 1.10.6
- microk8s v1.21.12 on ubuntu 21.10. Istio 1.10.6
- microk8s v1.21.12 on ubuntu 21.10. Istio 1.13.3

**Comment by howardjohn** [comment]
I can also reproduce pretty easily. Thanks for the detailed issue :+1:

**Comment by klarose** [comment]
No problem! I'm glad reproduction was easy. Hopefully this helps with that other issue as well, though I'm not sure it's the same unless multiple ports are involved on the server side. I was unable to reproduce this problem on pods which only exposed a single port. That makes sense given my hypothesis.

**Comment by howardjohn** [comment]
This is more workaround than fix since it requires NET_ADMIN on all pods, but the TPROXY mode (sidecar.istio.io/interceptionMode=TPROXY) annotation does fix this

**Comment by howardjohn** [comment]
I used https://github.com/howardjohn/istio/commit/75a67eb79c6ebf53850840f47cf4d93a52b46a00 to reproduce. Basically `client --flood svc:port1` then eventually `client svc:port2` fails.

Some ideas...
1. eBPF redirection. https://github.com/merbridge/merbridge/ might just automatically solve this today, untested
2. TPROXY - tested it works but requires NET_ADMIN
3. Redirect to multiple ports (eg 15006-15099). Costly in envoy today and just reduces likelyhood, doesn't fix it
4. Redirect to 127.0.0.{1..255} or similar instead of always 127.0.0.1. This would either be like (3) or could actually fix things if we have a dedicated mapping of dport -> daddress. But we don't know all the ports so its hard.
5. Try DNAT instead of REDIRECT. I have no reason to believe this would fix anything, but worth a shot in case there is some mechanism I am missing here

**Comment by klarose** [comment]
I wonder if there are any extensions to conntrack that allow one to extend the tuple in its table -- e.g. use a mark or something. If each redirect rule could also mark the request so that the kernel could properly distinguish between the two original ports, it may help.

**Comment by costinm** [comment]
I think the right fix for this is what we already planned and are implementing - remove interception and instead use CONNECT. 
Support is being added on client side (explicit DestinationRule initially, but can be done automatically), on server side we want to better support it for the (proposed) gateway improvements. It will also work on the sidecar. 

Egress is a different story: can this happen on egress ?  I suspect it could, so we may be back to the solutions above.

BTW, there is another one - may require NET_ADMIN on init only, create a TUN and use a soft TCP (there are a few), but perf is not great.

I'm shocked this was not found by others using REDIRECT.

**Comment by klarose** [comment]
@costinm I'm a bit surprised as well. When I first did this investigation I definitely focused on areas other than the sidecar first. I think what makes it a bit unlikely to happen in a normal case is that it requires that you do not enable retries for it to be truly visible, and further I suspect it only really becomes probable with long-lived connections.  Couple that with users typically building systems resilient to network failures, and even the occasional ones that occur for short-lived connections may have flown under the radar.

I only noticed because we started proxying a protocol that made it obvious when a connection failed.

**Comment by costinm** [comment]
Yes, typically redirect is used for 443/80, not for a range of ports
redirected to one. And retries and luck may hide the problem pretty well.

It is outbound traffic where this could be more frequent - since iptables
would overwrite both dest ip and port.

I'm still in disbelief that the kernel don't use the original dst ( which
is saved ) for TCP. Any chance something else - maybe in envoy - is causing
this?

**Comment by klarose** [comment]
I suppose something about how envoy accepts requests/some options on the listen socket could be at fault. That said, given that it's failing by incrementing the sync challenge counter, I'd be surprised if Envoy was even aware that the connection attempt had been made. FWIW, the relevant code in the kernel appears to be here: https://github.com/torvalds/linux/blob/master/net/ipv4/tcp_input.c#L5708

I haven't crawled through the code enough to understand what exactly influences the decision this layer of the stack makes.

**Comment by howardjohn** [comment]
FWIW I tried to see if we can avoid NET_ADMIN for TPROXY by using eBPF to set the IP_TRANSPARENT option; it almost works but the bpf has a limited set of `setsockopts` it allows and that isn't one of them.

**Comment by costinm** [comment]
If we use eBPF - we probably should go all in...

**Comment by donbowman** [comment]
This is still an active issue for us.

**Comment by howardjohn** [comment]
Would be interesting to see if ebpf solutions (https://istio.io/latest/blog/2022/merbridge/, etc) resolve this

**Comment by howardjohn** [comment]
One note: was browsing some eBPF network code, found https://github.com/cilium/cilium/blob/eddd2b3422cb8796900653595a267e8c7048c411/bpf/lib/proxy.h#L138. Suggests we could possibly unset source port when we set the dport and have it re-assigned in some cases

**Comment by luksa** [comment]
We at Red Hat have been investigating this issue for a few weeks and have found the root cause. When netfilter performs DNAT, it actually detects the collision between the two TCP connections and implicitly changes the source port of the second connection to avoid the collision. However, due to socket caching, it then doesn't distinguish between the two connections (apparently because the information was cached before the source port was changed). That's why it responds with an ACK on the first connection instead of a SYN-ACK on the second.

Fortunately, we can fix this issue by setting `net.ipv4.ip_early_demux=0` on the receiving end. 

I think we should allow users to set this and other values in `/proc/sys/` via istio-iptables. 

@costinm @howardjohn WDYT?

**Comment by howardjohn** [comment]
Nice find! Going to do some more research so I am up to speed but looks super promising. Looks like cilium does similar (https://docs.cilium.io/en/stable/operations/system_requirements/#requirements-for-l7-and-fqdn-policies) - they only do it without xt_socket, but I suspect that is for different reasons.

Operationally, this requires `privileged: true` to set, right? CNI, however, can also set it - so we may want to just allow it on CNI to simplify things

> I think we should allow users to set this and other values in /proc/sys/ via istio-iptables.

Not super sure we want to be a generalized sysctl injector. Any more context on this? I can see the benefits for future proofing just seems possibly out of scope.

**Comment by costinm** [comment]
To take a step back - any reason we want to keep going with REDIRECT ? 

AFAIK TPROXY doesn't have this problem - and it doesn't actually require extra permissions on interception side. We do require network permissions with our tproxy because it can spoof the source IP, but that's the second part and we can decouple it. 

I don't mind changing istio-cni to add the extra setting - but if we have to do it in istio-init and grant larger permissions I would just recommend switching to TPROXY as it is.

**Comment by howardjohn** [comment]
> We do require network permissions with our tproxy because it can spoof the source IP, but that's the second part and we can decouple it.

Can you use TPROXY without IP_TRANSPARENT if you don't do the source spoofing?

**Comment by luksa** [comment]
I've submitted a rough PR that adds the `sidecar.istio.io/disableIPEarlyDemux` pod annotation. When this is set to `true`, istio-cni sets `net.ipv4.ip_early_demux` to `0`.

**Comment by aojea** [comment]
> I've submitted a rough PR that adds the `sidecar.istio.io/disableIPEarlyDemux` pod annotation. When this is set to `true`, istio-cni sets `net.ipv4.ip_early_demux` to `0`.

@luksa I think that disabling the option will impact performance, do you have an estimation of that impact?

**Comment by luksa** [comment]
@aojea I don't have a clear answer, but disabling the option seems to _improve_ performance in some cases. This is from https://patchwork.ozlabs.org/project/netdev/patch/20120621235011.29846.29715.stgit@gitlad.jf.intel.com/:

> The main motivation behind this patch is to provide an option to disable
the [early demux] feature as it adds an additional cost to routing that reduces overall
throughput by up to 5%.  For example one of my systems went from 12.1Mpps
to 11.6 after the early socket demux was added.

This is also the reason why I think this should be set at the pod level instead of globally.

**Comment by aojea** [comment]
yeah, I saw that patch too ... but that is 2012, it will be good if we can reach out to some kernel developers for a more authoritative answer, when I was at RH I remember to work in a bug related with Florian Westphal

**Comment by costinm** [comment]
@howardjohn in context of ambient and the new istio-cni interception mechanisms - do we have an equivalent for this, and is the normal istio-cni setting ip_early_demux ?

**Comment by rjeberhard** [comment]
The TPROXY setting recommended by @howardjohn resolves this problem; however and unfortunately, this option is still marked as [alpha](https://istio.io/latest/docs/reference/config/annotations/#SidecarInterceptionMode).

Is there any plan to resolve this issue or what is the timeline for TPROXY support to move past the alpha stage?

**Comment by howardjohn** [comment]
Do to the excessive permissions required for TPROXY I don't consider it a particularly viable path forward

**Comment by wenlxie** [comment]
> We could see TcpExtTCPChallengeACK in the linux nstat tool incrementing when this occurs, which aligns with our theory.

`TCPSYNChallenge` should be more precise to check for this issue than `TcpExtTCPChallengeACK`

**Comment by wenlxie** [comment]
@klarose  @luksa
Does this issue also can be happened on the outbound?

**Comment by klarose** [comment]
> @klarose @luksa Does this issue also can be happened on the outbound?

I haven't seen it, but it certainly does seem possible -- I think the same redirects are configured in the opposite direction, right? As for why I haven't seen it, we tend to have a few orders of magnitude more inbound than outbound connections.

**Comment by wenlxie** [comment]
@klarose  Thanks for the reponse
I did some trace of the code. 
In the outbound direction, the  ip_early_demux feature will not be triggered because  skb->sk had been set when TCP start to transmit skb.

**Comment by luksa** [comment]
The netfilter issue that was causing this has apparently been fixed here: https://github.com/torvalds/linux/commit/e27c3295114bb6a6dc6d58a38f8503c0ea97aa6b

Can someone confirm?

**Comment by howardjohn** [comment]
> The netfilter issue that was causing this has apparently been fixed here: [torvalds/linux@e27c329](https://github.com/torvalds/linux/commit/e27c3295114bb6a6dc6d58a38f8503c0ea97aa6b)
> 
> Can someone confirm?

One nice way to confirm would be to remove our CI workarounds (we limit the concurrency), but the latest GKE is on 6.1 and the fix is in 6.7.

I have a more recent kernel locally though so will try to reproduce it in a bit

**Comment by sakatkoori** [comment]
@howardjohn @klarose , want to understand on when IGW creates a new connection and when it reuses the connection, if we see the problem here, when the connection is idle and the next connection is making use of the earlier source port at IGW is creating this issue, so, want to understand more on connection management of IGW. Could you please help understanding this?
